File: saral_hr/saral_hr/page/generate_monthly_loan_dues/generate_monthly_loan_dues.py
# Copyright (c) 2026, sj and contributors
# For license information, please see license.txt

import json

import frappe
from frappe.utils import flt, get_last_day, getdate

from saral_hr.saral_hr.doctype.employee_loan.employee_loan import month_sort_key, start_month_label

MONTHS = [
	"January",
	"February",
	"March",
	"April",
	"May",
	"June",
	"July",
	"August",
	"September",
	"October",
	"November",
	"December",
]


def _month_label(month, year):
	return f"{month} {year}"


def _period_dates(month, year):
	month_num = MONTHS.index(month) + 1
	start = getdate(f"{int(year)}-{month_num:02d}-01")
	return start, get_last_day(start)


def _period_locked(company, month, year):
	"""True when any submitted salary slip exists for this company in the month."""
	start, end = _period_dates(month, year)
	return bool(
		frappe.db.exists(
			"Salary Slip",
			{
				"company": company,
				"docstatus": 1,
				"start_date": ["between", [start, end]],
			},
		)
	)


def _list_dues(company, month_label, employee=None):
	filters = {"company": company, "month": month_label}
	if employee:
		filters["employee"] = employee
	rows = frappe.get_all(
		"Employee Loan Due",
		filters=filters,
		fields=[
			"name",
			"employee",
			"loan",
			"company",
			"month",
			"amount",
			"status",
			"salary_slip",
			"remarks",
		],
		order_by="employee asc, loan asc",
	)
	for row in rows:
		loan = frappe.db.get_value(
			"Employee Loan",
			row.loan,
			["full_name", "outstanding_amount", "expected_emi", "status"],
			as_dict=True,
		)
		row["full_name"] = loan.full_name if loan else ""
		row["outstanding_amount"] = loan.outstanding_amount if loan else 0
		row["expected_emi"] = loan.expected_emi if loan else 0
		row["loan_status"] = loan.status if loan else ""
		row["locked"] = 1 if row.salary_slip else 0
	return rows


def _create_missing_dues(company, month, year, employee=None):
	label = _month_label(month, year)
	target_key = month_sort_key(label)

	filters = {"docstatus": 1, "status": "Active", "company": company}
	if employee:
		filters["employee"] = employee

	created = 0
	skipped = 0
	loans = frappe.get_all(
		"Employee Loan",
		filters=filters,
		fields=[
			"name",
			"employee",
			"company",
			"expected_emi",
			"outstanding_amount",
			"start_month",
			"start_year",
		],
	)
	for loan in loans:
		start_label = start_month_label(loan.start_month, loan.start_year)
		if month_sort_key(start_label) > target_key:
			skipped += 1
			continue
		if flt(loan.outstanding_amount) <= 0:
			skipped += 1
			continue
		if frappe.db.exists("Employee Loan Due", {"loan": loan.name, "month": label}):
			skipped += 1
			continue
		amount = min(flt(loan.expected_emi), flt(loan.outstanding_amount))
		doc = frappe.get_doc(
			{
				"doctype": "Employee Loan Due",
				"employee": loan.employee,
				"loan": loan.name,
				"company": loan.company or company,
				"month": label,
				"year": int(year),
				"amount": amount,
				"status": "Skipped" if amount == 0 else "Pending",
			}
		)
		try:
			doc.insert(ignore_permissions=True)
			created += 1
		except (frappe.UniqueValidationError, frappe.DuplicateEntryError):
			skipped += 1
		except frappe.ValidationError as e:
			if "already exists" in str(e).lower():
				skipped += 1
			else:
				raise

	return created, skipped


@frappe.whitelist()
def open_period(company, month, year, employee=None):
	"""Single entry: load existing dues; create any missing when period is unlocked."""
	if not company or not month or not year:
		frappe.throw("Company, month and year are required.")
	if month not in MONTHS:
		frappe.throw(f"Invalid month: {month}")

	label = _month_label(month, year)
	locked = _period_locked(company, month, year)
	created = 0
	skipped = 0
	if not locked:
		created, skipped = _create_missing_dues(company, month, year, employee=employee)

	rows = _list_dues(company, label, employee=employee)
	if locked:
		for row in rows:
			row["locked"] = 1

	return {
		"rows": rows,
		"created": created,
		"skipped": skipped,
		"month": label,
		"period_locked": 1 if locked else 0,
	}


@frappe.whitelist()
def get_dues(company, month, year, employee=None):
	"""Backward-compatible read-only list (no create)."""
	if not company or not month or not year:
		frappe.throw("Company, month and year are required.")
	label = _month_label(month, year)
	locked = _period_locked(company, month, year)
	rows = _list_dues(company, label, employee=employee)
	if locked:
		for row in rows:
			row["locked"] = 1
	return rows


@frappe.whitelist()
def generate_dues(company, month, year, employee=None):
	"""Deprecated: use open_period. Kept for any old callers."""
	if _period_locked(company, month, year):
		return {"created": 0, "skipped": 0, "month": _month_label(month, year), "period_locked": 1}
	created, skipped = _create_missing_dues(company, month, year, employee=employee)
	return {"created": created, "skipped": skipped, "month": _month_label(month, year), "period_locked": 0}


@frappe.whitelist()
def save_dues(rows, company=None, month=None, year=None):
	if isinstance(rows, str):
		rows = json.loads(rows)

	if company and month and year and _period_locked(company, month, year):
		frappe.throw(
			f"Salary slips already exist for {_month_label(month, year)} — this period is locked."
		)

	updated = 0
	for row in rows or []:
		name = row.get("name")
		if not name:
			loan = row.get("loan")
			row_month = row.get("month")
			if not loan or not row_month:
				continue
			# Derive company/year from loan if period args missing
			if row_month and " " in row_month:
				parts = row_month.rsplit(" ", 1)
				if len(parts) == 2 and _period_locked(
					frappe.db.get_value("Employee Loan", loan, "company"),
					parts[0],
					parts[1],
				):
					frappe.throw(f"Salary slips already exist for {row_month} — this period is locked.")
			if frappe.db.exists("Employee Loan Due", {"loan": loan, "month": row_month}):
				frappe.throw(f"Due already exists for {loan} / {row_month}.")
			loan_doc = frappe.get_doc("Employee Loan", loan)
			doc = frappe.get_doc(
				{
					"doctype": "Employee Loan Due",
					"employee": loan_doc.employee,
					"loan": loan,
					"company": loan_doc.company,
					"month": row_month,
					"year": row.get("year"),
					"amount": flt(row.get("amount")),
					"remarks": row.get("remarks") or "",
				}
			)
			doc.insert(ignore_permissions=True)
			updated += 1
			continue

		doc = frappe.get_doc("Employee Loan Due", name)
		if doc.salary_slip and frappe.db.get_value("Salary Slip", doc.salary_slip, "docstatus") == 1:
			if abs(flt(doc.amount) - flt(row.get("amount"))) > 0.01:
				frappe.throw(f"Cannot change locked due {doc.loan} ({doc.month}).")
			continue
		doc.amount = flt(row.get("amount"))
		doc.remarks = row.get("remarks") or ""
		doc.save(ignore_permissions=True)
		updated += 1
	return {"updated": updated}


@frappe.whitelist()
def search_loans_for_due(company, month=None, year=None, employee=None):
	"""Active loans for Add Row picker. No month/due eligibility filtering."""
	if not company:
		frappe.throw("Company is required.")
	filters = {"docstatus": 1, "status": "Active", "company": company}
	if employee:
		filters["employee"] = employee

	loans = frappe.get_all(
		"Employee Loan",
		filters=filters,
		fields=[
			"name",
			"employee",
			"full_name",
			"amount",
			"expected_emi",
			"outstanding_amount",
			"start_month",
			"start_year",
			"loan_date",
			"reason",
			"proposed_tenure_months",
		],
		order_by="employee asc, name asc",
	)
	rows = []
	for loan in loans:
		suggested = min(flt(loan.expected_emi), flt(loan.outstanding_amount)) if flt(loan.outstanding_amount) > 0 else flt(loan.expected_emi)
		reason = (loan.reason or "").strip().replace("\n", " ")
		if len(reason) > 80:
			reason = reason[:77] + "…"
		rows.append(
			{
				"name": loan.name,
				"employee": loan.employee,
				"full_name": loan.full_name,
				"amount": flt(loan.amount),
				"expected_emi": flt(loan.expected_emi),
				"outstanding_amount": flt(loan.outstanding_amount),
				"suggested_amount": suggested,
				"start": f"{loan.start_month} {loan.start_year}",
				"loan_date": loan.loan_date,
				"reason": reason,
				"proposed_tenure_months": loan.proposed_tenure_months,
				"label": (
					f"{loan.employee} · {loan.full_name or ''} · {loan.name} · "
					f"Outst {flt(loan.outstanding_amount):,.0f} · EMI {flt(loan.expected_emi):,.0f}"
				),
			}
		)
	return rows

File: saral_hr/saral_hr/page/generate_monthly_loan_dues/generate_monthly_loan_dues.js
frappe.pages["generate-monthly-loan-dues"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Generate Monthly Loan Dues"),
		single_column: true,
	});
	wrapper._gld_page = page;
};

frappe.pages["generate-monthly-loan-dues"].on_page_show = function (wrapper) {
	$("body").removeClass("full-width");
	const $nb = $("#navbar-breadcrumbs");
	if ($nb.length) {
		$nb.empty().append(
			`<li><a href="/app/saral-hr">${__("Saral HR")}</a></li>`,
			`<li><a href="/app/generate-monthly-loan-dues">${__("Generate Monthly Loan Dues")}</a></li>`
		);
	}
	document.title = __("Generate Monthly Loan Dues");
	if (wrapper._gld_page) {
		wrapper._gld_page.set_title(__("Generate Monthly Loan Dues"));
	}
	const $main = $(wrapper).find(".layout-main-section");
	$main.html(gld_shell_html());
	init_generate_loan_dues($main, wrapper._gld_page);
};

function gld_shell_html() {
	const years = [];
	const cy = new Date().getFullYear();
	// Cover historical payroll months (loans often start in prior FY).
	for (let y = cy - 3; y <= cy + 2; y++) years.push(String(y));
	const months = [
		"January","February","March","April","May","June",
		"July","August","September","October","November","December"
	];
	const now = new Date();
	return `
	<div class="gld-root" style="padding:8px 0 40px;">
		<div class="frappe-card" style="padding:14px 16px;margin-bottom:12px;">
			<div style="display:flex;flex-wrap:wrap;gap:12px;align-items:flex-end;">
				<div>
					<label class="control-label">${__("Company")}</label>
					<div class="gld-company"></div>
				</div>
				<div>
					<label class="control-label">${__("Month")}</label>
					<select class="form-control gld-month" style="width:140px;">
						${months.map((m,i)=>`<option value="${m}" ${i===now.getMonth()?"selected":""}>${m}</option>`).join("")}
					</select>
				</div>
				<div>
					<label class="control-label">${__("Year")}</label>
					<select class="form-control gld-year" style="width:100px;">
						${years.map(y=>`<option value="${y}" ${y===String(cy)?"selected":""}>${y}</option>`).join("")}
					</select>
				</div>
				<div>
					<label class="control-label">${__("Employee")}</label>
					<div class="gld-employee"></div>
				</div>
				<button class="btn btn-default btn-sm gld-load">${__("Load")}</button>
				<button class="btn btn-primary btn-sm gld-generate">${__("Generate")}</button>
				<button class="btn btn-success btn-sm gld-save">${__("Save")}</button>
				<button class="btn btn-default btn-sm gld-add">${__("Add Row")}</button>
			</div>
		</div>
		<div class="frappe-card gld-table-wrap" style="padding:12px;overflow:auto;">
			<div class="text-muted">${__("Select company and month, then Generate or Load.")}</div>
		</div>
	</div>`;
}

function init_generate_loan_dues($main, page) {
	const state = { rows: [], generating: false };

	const company = frappe.ui.form.make_control({
		parent: $main.find(".gld-company"),
		df: {
			fieldtype: "Link",
			options: "Company",
			fieldname: "company",
			placeholder: __("Company"),
			default: frappe.defaults.get_user_default("Company"),
		},
		render_input: true,
	});
	company.refresh();

	const employee = frappe.ui.form.make_control({
		parent: $main.find(".gld-employee"),
		df: {
			fieldtype: "Link",
			options: "Company Link",
			fieldname: "employee",
			placeholder: __("Optional"),
		},
		render_input: true,
	});
	employee.refresh();

	function filters() {
		return {
			company: company.get_value(),
			month: $main.find(".gld-month").val(),
			year: $main.find(".gld-year").val(),
			employee: employee.get_value() || "",
		};
	}

	function render() {
		const $wrap = $main.find(".gld-table-wrap");
		if (!state.rows.length) {
			$wrap.html(`<div class="text-muted">${__("No dues for this period.")}</div>`);
			return;
		}
		let html = `<table class="table table-bordered" style="margin:0;">
			<thead><tr>
				<th>${__("Employee")}</th>
				<th>${__("Name")}</th>
				<th>${__("Loan")}</th>
				<th style="text-align:right;">${__("Outstanding")}</th>
				<th style="text-align:right;">${__("Amount")}</th>
				<th>${__("Status")}</th>
				<th>${__("Salary Slip")}</th>
				<th>${__("Remarks")}</th>
			</tr></thead><tbody>`;
		state.rows.forEach((r, idx) => {
			const locked = !!r.salary_slip;
			html += `<tr data-idx="${idx}">
				<td>${frappe.utils.escape_html(r.employee || "")}</td>
				<td>${frappe.utils.escape_html(r.full_name || "")}</td>
				<td><a href="/app/employee-loan/${encodeURIComponent(r.loan)}">${frappe.utils.escape_html(r.loan || "")}</a></td>
				<td style="text-align:right;">${format_currency(r.outstanding_amount || 0)}</td>
				<td style="text-align:right;">
					<input type="number" step="0.01" class="form-control input-sm gld-amount"
						value="${flt(r.amount)}" ${locked ? "disabled" : ""} style="max-width:120px;margin-left:auto;" />
				</td>
				<td>${frappe.utils.escape_html(r.status || "")}</td>
				<td>${r.salary_slip ? `<a href="/app/salary-slip/${encodeURIComponent(r.salary_slip)}">${frappe.utils.escape_html(r.salary_slip)}</a>` : "—"}</td>
				<td><input type="text" class="form-control input-sm gld-remarks" value="${frappe.utils.escape_html(r.remarks || "")}" ${locked ? "disabled" : ""} /></td>
			</tr>`;
		});
		html += "</tbody></table>";
		$wrap.html(html);
	}

	function collect() {
		$main.find("tbody tr").each(function () {
			const idx = cint($(this).attr("data-idx"));
			if (state.rows[idx]) {
				state.rows[idx].amount = flt($(this).find(".gld-amount").val());
				state.rows[idx].remarks = $(this).find(".gld-remarks").val();
			}
		});
		return state.rows;
	}

	function load_dues() {
		const f = filters();
		if (!f.company) return frappe.msgprint(__("Select Company"));
		frappe.call({
			method: "saral_hr.saral_hr.page.generate_monthly_loan_dues.generate_monthly_loan_dues.get_dues",
			args: f,
			freeze: true,
			callback(r) {
				state.rows = r.message || [];
				render();
			},
		});
	}

	// Replacing HTML on each show leaves delegated handlers on $main stacked —
	// a second Generate click fired in parallel and raced past the exists check.
	$main.off(".gld");

	$main.on("click.gld", ".gld-load", () => load_dues());

	$main.on("click.gld", ".gld-generate", () => {
		const f = filters();
		if (!f.company) return frappe.msgprint(__("Select Company"));
		if (state.generating) return;
		state.generating = true;
		frappe.call({
			method: "saral_hr.saral_hr.page.generate_monthly_loan_dues.generate_monthly_loan_dues.generate_dues",
			args: f,
			freeze: true,
			freeze_message: __("Generating…"),
			callback(r) {
				state.generating = false;
				frappe.show_alert({
					message: __("Created {0}, skipped {1}", [r.message.created, r.message.skipped]),
					indicator: "green",
				});
				load_dues();
			},
			error() {
				state.generating = false;
			},
		});
	});

	$main.on("click.gld", ".gld-save", () => {
		const rows = collect();
		frappe.call({
			method: "saral_hr.saral_hr.page.generate_monthly_loan_dues.generate_monthly_loan_dues.save_dues",
			args: { rows },
			freeze: true,
			callback(r) {
				frappe.show_alert({
					message: __("Saved {0} row(s)", [r.message.updated]),
					indicator: "green",
				});
				load_dues();
			},
		});
	});

	$main.on("click.gld", ".gld-add", () => {
		const f = filters();
		if (!f.company) return frappe.msgprint(__("Select Company"));
		const d = new frappe.ui.Dialog({
			title: __("Add Loan Due"),
			fields: [
				{ fieldname: "loan", fieldtype: "Link", options: "Employee Loan", label: __("Loan"), reqd: 1,
					get_query: () => ({ filters: { docstatus: 1, status: "Active", company: f.company } }) },
				{ fieldname: "amount", fieldtype: "Currency", label: __("Amount"), reqd: 1 },
				{ fieldname: "remarks", fieldtype: "Small Text", label: __("Remarks") },
			],
			primary_action_label: __("Add"),
			primary_action(values) {
				const month = `${f.month} ${f.year}`;
				frappe.call({
					method: "saral_hr.saral_hr.page.generate_monthly_loan_dues.generate_monthly_loan_dues.save_dues",
					args: {
						rows: [{
							loan: values.loan,
							month,
							year: cint(f.year),
							amount: values.amount,
							remarks: values.remarks || "",
						}],
					},
					freeze: true,
					callback() {
						d.hide();
						load_dues();
					},
				});
			},
		});
		d.show();
	});
}
